### ISSUE TITLE
[jvm-packages] xgboost4j-spark XGBoostModel.eval doesn't use provided eval_metric

### ISSUE THREAD

**Issue by ebernhardson** [dmlc/xgboost#2239]
For bugs or installation issues, please provide the following information.
The more information you provide, the more easily we will be able to offer
help and advice.

## Environment info
Operating System:
ubuntu trusty

Compiler:
g++4.8, openjdk 1.8.0_111

Package used (python/R/jvm/C++):
xgboost4j-spark

`xgboost` version used:
0.7

If installing from source, please provide 

1. The commit hash (`git rev-parse HEAD`)
2. Logs will be helpful (If logs are large, please upload as attachment).

## Steps to reproduce

1. Load spark-shell with xgboost4j-spark jar:
~/spark-2.1.0-bin-hadoop2.6/bin/spark-shell --jars ~/xgboost4j-spark-0.7-jar-with-dependencies.jar
2. Paste the following:
```
import org.apache.spark.ml.feature.LabeledPoint
import org.apache.spark.ml.linalg.Vectors

val train = sc.parallelize(Seq(LabeledPoint(1.0, Vectors.dense(2.2, 4.4)), LabeledPoint(2.0, Vectors.dense(4.4, 2.2))))

val params = Map(("objective", "rank:pairwise"), ("eval_metric", "ndcg"), ("groupData", Seq(Seq(2))))
val model = ml.dmlc.xgboost4j.scala.spark.XGBoost.trainWithRDD(train, params, 1, 1)
model.eval(train, "train", null, 0)
```

Resulting output:
```
sing Scala version 2.11.8 (OpenJDK 64-Bit Server VM, Java 1.8.0_111)
Type in expressions to have them evaluated.
Type :help for more information.

scala> import org.apache.spark.ml.feature.LabeledPoint
import org.apache.spark.ml.feature.LabeledPoint

scala> import org.apache.spark.ml.linalg.Vectors
import org.apache.spark.ml.linalg.Vectors

scala> val train = sc.parallelize(Seq(LabeledPoint(1.0, Vectors.dense(2.2, 4.4)), LabeledPoint(2.0, Vectors.dense(4.4, 2.2))))
train: org.apache.spark.rdd.RDD[org.apache.spark.ml.feature.LabeledPoint] = ParallelCollectionRDD[0] at parallelize at <console>:26

scala> val params = Map(("objective", "rank:pairwise"), ("eval_metric", "ndcg"), ("groupData", Seq(Seq(2))))
params: scala.collection.immutable.Map[String,Object] = Map(objective -> rank:pairwise, eval_metric -> ndcg, groupData -> List(List(2)))

scala> val model = ml.dmlc.xgboost4j.scala.spark.XGBoost.trainWithRDD(train, params, 1, 1)
Tracker started, with env={DMLC_TRACKER_URI=10.64.5.104, DMLC_NUM_WORKER=1, DMLC_TRACKER_PORT=9096, DMLC_NUM_SERVER=0}
[Stage 0:>                                                        (0 + 16) / 16][01:13:43] src/tree/updater_prune.cc:74: tree pruning end, 1 roots, 2 extra nodes, 0 pruned nodes, max_depth=1
model: ml.dmlc.xgboost4j.scala.spark.XGBoostModel = XGBoostRegressionModel_7a977c219162

scala> model.eval(train, "train", null, 0)
res0: String = [0]      train-map = 1.0
```

I'm expecting that final line to be something like [0]    train-ndcg = 1.0, but instead it's returning the default objective metric of rank:pairwise. This also happens with full size examples, but the above is a minimal reproduction case.

## What have you tried?

1. recompiled with master branch, commit 8ab5d461, same problem

**Comment by CodingCat** [comment]
there is a bug when serializing learner, in which the metrics are not saved https://github.com/dmlc/xgboost/blob/master/src/learner.cc#L285

**Comment by ebernhardson** [comment]
Thanks for the pointer, I've pushed up a patch to fix that and verified i can now eval with the chosen metric in xgboost4j-spark.

**Comment by CodingCat** [comment]
fixed by https://github.com/dmlc/xgboost/pull/2241
